File: SimpleEngine/StringSeperator.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace SimpleEngine
{
	public class StringSeperator
	{
		public StringSeperator(string aSeperator, string aFinalSeperator, string empty)
		{
			mSeperator = aSeperator;
			mFinalSeperator = aFinalSeperator;
			mEmpty = empty;
		}
		public StringSeperator(string aSeperator, string aFinalSeperator)
		{
			mSeperator = aSeperator;
			mFinalSeperator = aFinalSeperator;
			mEmpty = "";
		}
		public StringSeperator(string aSeperator)
		{
			mSeperator = aSeperator;
			mFinalSeperator = aSeperator;
		}

		public StringSeperator Append(string astring)
		{
			mstrings.Add(astring);
			return this;
		}

		public StringSeperator Append(IEnumerable<string> astrings)
		{
			mstrings.AddRange(astrings);
			return this;
		}
		public StringSeperator Append(object[] d)
		{
			foreach (object o in d)
				mstrings.Add(o.ToString());
			return this;
		}

		public StringSeperator AppendFormat(string aFormat, params Object[] aValues)
		{
			StringBuilder builder = new StringBuilder();
			builder.AppendFormat(aFormat, aValues);
			Append(builder.ToString());
			return this;
		}

		public void Clear()
		{
			mstrings.Clear();
		}

		public override string ToString()
		{
			if (mstrings.Count == 0) return mEmpty;
			StringBuilder builder = new StringBuilder();
			for (int index = 0; index < mstrings.Count; ++index)
			{
				string value = mstrings[index];
				builder.Append(value);

				if (mstrings.Count != index + 1) // if this item isnt the last one in the list
				{
					string seperator = mSeperator;
					if (mstrings.Count == index + 2)
					{
						seperator = mFinalSeperator;
					}
					builder.Append(seperator);
				}
			}
			return builder.ToString();
		}

		List<string> mstrings = new List<string>();
		readonly string mSeperator;
		readonly string mFinalSeperator;
		readonly string mEmpty;
	}
}

File: src/engine/stringseperator.cc
#include "engine/stringseperator.h"

namespace simple
{
    StringSeperator::StringSeperator(const std::string& s, const std::string& fs, const std::string& e)
        : seperator(s)
        , final_seperator(fs)
        , empty(e)
    {
    }

    StringSeperator::StringSeperator(const std::string& s, const std::string& fs)
        : seperator(s)
        , final_seperator(fs)
        , empty("")
    {
    }

    StringSeperator::StringSeperator(const std::string& s)
        : seperator(s)
        , final_seperator(s)
    {
    }

    std::string StringSeperator::ToString(const std::vector<std::string>& strings) const
    {
        if (strings.empty())
            return empty;

        std::ostringstream builder;
        for (int index = 0; index < strings.size(); ++index)
        {
            const std::string& value = strings[index];
            builder << value;

            if (strings.size() != index + 1)  // if this item isnt the last one: the list
            {
                const std::string the_seperator =
                    (strings.size() == index + 2) ? final_seperator
                                                  : seperator;
                builder << the_seperator;
            }
        }
        return builder.str();
    }
}
